Index: src/config_manager.cc
===================================================================
--- src/config_manager.cc	(revision 2018)
+++ src/config_manager.cc	(working copy)
@@ -628,6 +628,14 @@
 #ifdef EXTERNAL_TRANSCODING
     config->appendElementChild(renderTranscodingSection());
 #endif
+
+#ifdef HAVE_LASTFMLIB
+    Ref<Element> lastfm(new Element(_("lastfm")));
+    lastfm->setAttribute(_("enabled"), _(DEFAULT_LASTFM_ENABLED));
+    lastfm->appendTextChild(_("username"), _(DEFAULT_LASTFM_USERNAME));
+    lastfm->appendTextChild(_("password"), _(DEFAULT_LASTFM_PASSWORD));
+    config->appendElementChild(lastfm);
+#endif
     
     config->indent();
     save_text(config_filename, config->print());
@@ -1999,6 +2007,42 @@
     SET_BOOL_OPTION(CFG_ONLINE_CONTENT_WEBORAMA_UPDATE_AT_START);
 #endif
 
+#if defined(HAVE_LASTFMLIB)
+    temp = getOption(_("/lastfm/attribute::enabled"), _(DEFAULT_LASTFM_ENABLED));
+
+    if (!validateYesNo(temp))
+        throw _Exception(_("Error in config file: "
+                           "invalid \"enabled\" attribute value in "
+                           "<lastfm> tag"));
+
+    NEW_BOOL_OPTION(temp == "yes" ? true : false);
+    SET_BOOL_OPTION(CFG_LASTFM_ENABLED);
+
+    if (temp == YES)
+    {
+        temp = getOption(_("/lastfm/username"), _(DEFAULT_LASTFM_USERNAME));
+
+        if (temp.length() == 0)
+            throw _Exception(_("Error in config file: lastfm - "
+                               "invalid username value in "
+                               "<username> tag"));
+
+        NEW_OPTION(temp);
+        SET_OPTION(CFG_LASTFM_USERNAME);
+
+        temp = getOption(_("/lastfm/password"), _(DEFAULT_LASTFM_PASSWORD));
+
+        if (temp.length() == 0)
+            throw _Exception(_("Error in config file: lastfm - "
+                               "invalid password value in "
+                               "<password> tag"));
+
+        NEW_OPTION(temp);
+        SET_OPTION(CFG_LASTFM_PASSWORD);
+    }
+#endif
+
+
 #ifdef ATRAILERS
     temp = getOption(_("/import/online-content/AppleTrailers/attribute::enabled"), 
                      _(DEFAULT_ATRAILERS_ENABLED));
Index: src/server.cc
===================================================================
--- src/server.cc	(revision 2018)
+++ src/server.cc	(working copy)
@@ -37,6 +37,10 @@
     #include <curl/curl.h>
 #endif
 
+#ifdef HAVE_LASTFMLIB
+    #include "lastfm.h"
+#endif
+
 #include "server.h"
 #include "web_callbacks.h"
 #include "content_manager.h"
@@ -100,6 +104,10 @@
 #ifdef HAVE_CURL
     curl_global_init(CURL_GLOBAL_ALL);
 #endif
+
+#ifdef HAVE_LASTFMLIB
+    LastFm::getInstance()->initialize();
+#endif
 }
 
 void Server::upnp_init(String iface, String ip_address, int port)
@@ -334,6 +342,10 @@
     curl_global_cleanup();
 #endif
 
+#ifdef HAVE_LASTFMLIB
+    LastFm::getInstance()->destroy();
+#endif
+
     log_debug("now calling upnp finish\n");
     UpnpFinish();
     storage = nil;
Index: src/common.h
===================================================================
--- src/common.h	(revision 2018)
+++ src/common.h	(working copy)
@@ -352,6 +352,12 @@
     #define DEFAULT_FFMPEGTHUMBNAILER_WORKAROUND_BUGS   NO
 #endif
 
+#if defined(HAVE_LASTFMLIB)
+    #define DEFAULT_LASTFM_ENABLED  NO
+    #define DEFAULT_LASTFM_USERNAME "lastfmuser"
+    #define DEFAULT_LASTFM_PASSWORD "lastfmpass"
+#endif
+
 #define DEFAULT_MARK_PLAYED_ITEMS_ENABLED               NO
 #define DEFAULT_MARK_PLAYED_ITEMS_SUPPRESS_CDS_UPDATES   YES
 #define DEFAULT_MARK_PLAYED_ITEMS_STRING_MODE           "prepend"
Index: src/config_manager.h
===================================================================
--- src/config_manager.h	(revision 2018)
+++ src/config_manager.h	(working copy)
@@ -187,6 +187,12 @@
     CFG_ONLINE_CONTENT_ATRAILERS_PURGE_AFTER,
     CFG_ONLINE_CONTENT_ATRAILERS_RESOLUTION,
 #endif
+#ifdef HAVE_LASTFMLIB
+    CFG_LASTFM_ENABLED,
+    CFG_LASTFM_USERNAME,
+    CFG_LASTFM_PASSWORD,
+#endif
+
     CFG_MAX
 } config_option_t;
 
Index: src/file_request_handler.cc
===================================================================
--- src/file_request_handler.cc	(revision 2018)
+++ src/file_request_handler.cc	(working copy)
@@ -66,6 +66,10 @@
     #include "transcoding/transcode_dispatcher.h"
 #endif
 
+#ifdef HAVE_LASTFMLIB
+    #include "lastfm.h"
+#endif
+
 using namespace zmm;
 using namespace mxml;
 
@@ -452,8 +456,15 @@
     }
 
     Ref<CdsItem> item = RefCast(obj, CdsItem);
-
+    
     String path = item->getLocation();
+    
+#ifdef HAVE_LASTFMLIB
+    if (item->getMimeType().startsWith(String("audio")))
+    {
+        LastFm::getInstance()->startedPlaying(item);
+    }
+#endif
 
     String ext = dict->get(_("ext"));
     int edot = ext.rindex('.');
Index: build/libmediatomb_src
===================================================================
--- build/libmediatomb_src	(revision 2018)
+++ build/libmediatomb_src	(working copy)
@@ -67,6 +67,8 @@
 ../src/layout/js_layout.cc \
 ../src/layout/js_layout.h \
 ../src/layout/layout.h \
+../src/lastfm.cc \
+../src/lastfm.h \
 ../src/logger.cc \
 ../src/logger.h \
 ../src/main.cc \
Index: build/Makefile.am
===================================================================
--- build/Makefile.am	(revision 2018)
+++ build/Makefile.am	(working copy)
@@ -27,7 +27,8 @@
                      $(CURL_CFLAGS) \
                      $(EXPAT_CFLAGS) \
                      $(LIBMP4V2_CFLAGS) \
-                     $(LIBDVDNAV_CFLAGS)
+                     $(LIBDVDREAD_CFLAGS) \
+                     $(LASTFMLIB_CFLAGS)
 
 bin_PROGRAMS = mediatomb
 mediatomb_SOURCES = $(top_srcdir)/src/main.cc
@@ -53,7 +54,9 @@
                      $(CURL_CFLAGS) \
                      $(EXPAT_CFLAGS) \
                      $(LIBMP4V2_CFLAGS) \
-                     $(LIBDVDNAV_CFLAGS)
+                     $(LIBDVDREAD_CFLAGS) \
+                     $(LIBDVDNAV_CFLAGS) \
+					 $(LASTFMLIB_CFLAGS)
 
 mediatomb_LDADD = \
     libmediatomb.a \
@@ -90,5 +93,6 @@
     $(NSL_LIBS) \
     $(LWRES_LIBS) \
     $(LIBDVDNAV_LIBS) \
-    $(CURL_LIBS)
-    
+    $(CURL_LIBS)\
+    $(LASTFMLIB_LIBS)
+
Index: configure.ac
===================================================================
--- configure.ac	(revision 2018)
+++ configure.ac	(working copy)
@@ -1569,6 +1569,12 @@
 LDFLAGS="$LDFLAGS_SAVE"
 CPPFLAGS="$CPPFLAGS_SAVE"
 
+######## lastfm
+MT_CHECK_OPTIONAL_PACKAGE([lastfmlib], [disable],
+        [compile with lastfmlib support for Last.fm scrobbling],
+        [lastfmlib/lastfmscrobbler],
+        [lastfm], [create_scrobbler])
+
 ######## extractor
 
 if test "x$LIBEXTRACTOR_OPTION_ENABLED" = xyes; then
@@ -1899,6 +1905,7 @@
 echo "libdvdnav             : $LIBDVDNAV_STATUS"
 echo "ffmpeg                : $FFMPEG_STATUS"
 echo "ffmpegthumbnailer     : $FFMPEGTHUMBNAILER_STATUS"
+echo "lastfm                : $LASTFMLIB_STATUS"
 echo "external transcoding  : $EXTERNAL_TRANSCODING_OPTION_ENABLED"
 echo "curl                  : $CURL_OK"
 echo "YouTube               : $YOUTUBE_OPTION_ENABLED"
